snowflake/src/main/java/org/apache/iceberg/snowflake/JdbcSnowflakeClient.java

```diff
@@ -0,0 +1,163 @@
+          "Failed to list schemas for namespace %s",
+          namespace != null ? namespace.toString() : "");
+    } catch (InterruptedException e) {
+      throw new UncheckedInterruptedException(e, "Interrupted while listing schemas");
+    }
+    return schemas;
+  }
+
+  @Override
+  public List<SnowflakeTable> listIcebergTables(Namespace namespace) {
+    StringBuilder baseQuery = new StringBuilder("SHOW ICEBERG TABLES");
+    Object[] queryParams = null;
+    if (namespace.length() == SnowflakeResources.MAX_NAMESPACE_DEPTH) {
+      // For two level namespace, search for iceberg tables within the given schema.
+      baseQuery.append(" IN SCHEMA IDENTIFIER(?)");
+      queryParams =
+          new Object[] {
+            String.format(
+                "%s.%s",
+                namespace.level(SnowflakeResources.NAMESPACE_DB_LEVEL - 1),
+                namespace.level(SnowflakeResources.NAMESPACE_SCHEMA_LEVEL - 1))
+          };
+    } else if (namespace.length() == SnowflakeResources.NAMESPACE_DB_LEVEL) {
+      // For one level namespace, search for iceberg tables within the given database.
+      baseQuery.append(" IN DATABASE IDENTIFIER(?)");
+      queryParams = new Object[] {namespace.level(SnowflakeResources.NAMESPACE_DB_LEVEL - 1)};
+    } else {
+      // For empty or db level namespace, search at account level.
+      baseQuery.append(" IN ACCOUNT");
+    }
+
+    final String finalQuery = baseQuery.toString();
+    final Object[] finalQueryParams = queryParams;
+    List<SnowflakeTable> tables;
+    try {
+      tables =
+          connectionPool.run(
+              conn ->
+                  queryRunner.query(
+                      conn, finalQuery, SnowflakeTable.createHandler(), finalQueryParams));
+    } catch (SQLException e) {
+      throw new UncheckedSQLException(
+          e, "Failed to list tables for namespace %s", namespace.toString());
+    } catch (InterruptedException e) {
+      throw new UncheckedInterruptedException(e, "Interrupted while listing tables");
+    }
+    return tables;
+  }
+
+  @Override
+  public SnowflakeTableMetadata getTableMetadata(TableIdentifier tableIdentifier) {
+    SnowflakeTableMetadata tableMeta;
+    try {
+      final String finalQuery = "SELECT SYSTEM$GET_ICEBERG_TABLE_INFORMATION(?) AS METADATA";
+      tableMeta =
+          connectionPool.run(
+              conn ->
+                  queryRunner.query(
+                      conn,
+                      finalQuery,
+                      SnowflakeTableMetadata.createHandler(),
+                      tableIdentifier.toString()));
+    } catch (SQLException e) {
+      throw new UncheckedSQLException(
+          e, "Failed to get table metadata for %s", tableIdentifier.toString());
+    } catch (InterruptedException e) {
+      throw new UncheckedInterruptedException(e, "Interrupted while getting table metadata");
+    }
+    return tableMeta;
+  }
+
+  @Override
+  public void close() {
+    connectionPool.close();
```

Comment:
should this have a null check?

Reply:
Decided on null check in the constructor instead as a checkArgument.